compiler/rustc_span/src/hygiene.rs

```diff
@@ -614,6 +617,31 @@ impl HygieneData {
         };
         ctxt
     }
+
+    /// Helper to allocate context only if it would be different from the current one.
+    /// This avoids unnecessary recursive calls in common cases.
```

Comment:
Technically we do not avoid a recursive call :D
Previously we recursively called `alloc_ctxt` and immediately returned in the cache check.
Now we recursively call `alloc_ctxt_if_needed` and immediately return in the `parent.0 as usize == current_idx` check.

So the only real optimization in the `alloc_ctxt` check is replacing a hash map lookup `syntax_context_map.get(...)` with a cheap arithmetic check.
Although I still need to convince myself that this cheaper check is correct.